### PR TITLE
[Accessibility]: move blame info from gutter into CodeMirror decorations to make them readable by screenreaders

### PR DESCRIPTION
**Pull request by taras-yemets**
Closes https://github.com/sourcegraph/sourcegraph/issues/43211

Moves blame decorations from gutter to the code line.
Two-dimensional code navigation suggestion suggested in [this comment](https://github.com/sourcegraph/sourcegraph/pull/44148#issuecomment-1311160200) will be addressed in a follow-up issue https://github.com/sourcegraph/sourcegraph/issues/44594.

https://user-images.githubusercontent.com/25318659/202664202-25398d5d-aa42-49b4-bb28-3443833450d6.mov

⚠️ **Implementation details**
1. Blame decorations were moved from a separate div inside the line with the code and no longer had full-height background. I tried to achieve full-height-like view of inline decorations with CSS only by playing around with spacing and pseudo elements ([what we do with the default blob viewer](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@2d8f7d216cc4ce7e0f05fd64d7a1bead9c3d98f4/-/blob/client/web/src/repo/blob/Blob.tsx?L798-831) but with `::before` and `::after`), but it neither approach didn't work: spacing increases code line hight and it looks weird when selected/highlighted and pseudo elements are removed by the CodeMirror editor.
To achieve the full-height-like column view with inline blame decorations we render empty gutter providing with and background color and shift the line content to the left by `var(--blame-decoration-width)` so that inline blame decorations make use of the full-height background provided by the gutter.
This approach comes with a drawback: we have to [disable gutters sticky position](https://github.com/sourcegraph/sourcegraph/pull/44148/files#diff-87e828a8ef5397d2e106d753e3616e665592ac5d3bc2624373dc71426414d994R189-R191) (including line numbers) so that the blame gutter background is not scrolled horizontally with the content while the inline blame decoration is not.
2. The suggested approach improves git blame decorations readability (they become accessible by the screenreader and can be read line by line along with the code they reference to). But it looks a bit hacky as we still use two different approaches (inline decorations and gutter) to achieve this effect and it may be not the way the CodeMirror authors expect us to use the editor. CodeMirror on its website claims that it "works well with screen readers and keyboard-only users" so I'll browse their forum to find similar use-cases or post an issue to get some guidance on how to make gutters keyboard and screenreader accessible. If there's a better approach we'll address it in a follow-up PR. cc: @fkling 

**Previous description 👇🏻**  (didn't remove as it's referenced in comments, please ignore when reviewing the suggested changes)
**>> start**
Removes `aria-hidden="false"` from the CodeMirror editor gutters.

Although it fixes the referenced issue, I'm not sure it is the best approach to the CodeMirror file view with blame decorations open readability.
When the screen reader reaches the code editor area, it reads the content in the following order (see screenshot):
1) all the line numbers one by one
2) all the blame decorations one by one 
3) code content.

Screenshot 2022-11-09 at 15 44 16

And it doesn't look like a perfect user experience 😞 

Although we can hide line numbers from the screen reader (but do we really want it: Monaco editor and GitHub blame view, for example, announce line numbers), it doesn't help with reading all the blame decorations one by one before reaching the file content.

**Potential solution**
Line numbers and other types of content CodeMirror users may want to put into gutters ([gutter description](https://codemirror.net/docs/ref/#h_gutters)) are not expected to be accessible by the CodeMirror creators  ([source](https://sourcegraph.com/github.com/codemirror/view@e1b25c4c5444efbf94eb745db1d485fdc7f624a7/-/blob/src/gutter.ts?L122)). 
The potential solution could be to move blame decorations (and maybe line numbers) inside the line content (e.g., prepend the line code content with line number and commit info - see screenshot) so that they are announced in a more consumable way (e.g., "line 1", "$COMMIT_INFO", "$LINE_CONTENT"). But I'm not sure if it's a feasible way as I'm not a power CodeMirror user (I hope @fkling can help here).

Screenshot 2022-11-09 at 16 09 43

**>> end**

## Test plan
Tested manually that the git blame column content is read by the screenreader (VoiceOver). 

## App preview:

- [Web](https://sg-web-ps-make-codemirror-gutters.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.00% (+0.10 kb) | 0.00% (+0.10 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/ps/make-codemirror-gutters-readable-by-screenreaders/compare-report.html;tab=live_object) for a full comparison between the commits d47ab6c and d0b1412 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Comment by philipp-spiess** [comment]
Thanks for the writeup @taras-yemets.

> The potential solution could be to move blame decorations (and maybe line numbers) inside the line content (e.g., prepend the line code content with line number and commit info - see screenshot) so that they are announced in a more consumable way (e.g., "line 1", "$COMMIT_INFO", "$LINE_CONTENT"). But I'm not sure if it's a feasible way as I'm not a power CodeMirror user (I hope @fkling can help here).

Yeah if this would be possible this would be the best solution.

---

I think another problem with announcing line numbers with this DOM structure is that the screenreader would first read all line numbers one by one (1, 2, 3, ...) before it even gets to the code part. So this does not seem very useful :/ So maybe the line numbers should stay `aria-hidden="true"`? 

The disconnect between the git blame annotation and the line of code is also annoying but maybe we can add a screenreader-only line number in here (so we only announce line numbers if we render git blame). This way there is at least some connection between the git blame information and the line of code? E.g.:

```diff
  <div class="cm-gutter BlameColumn">
      <div class="BlameRow">
+       <div class="sr_only">123</div>
        ...the rest...
       </div>
     </div>
 </div>
```

**Comment by fkling** [comment]
How is the code read? Line by line? It would be possible to insert a screenreader only widget decoration at the start of each line.

File: client/web/src/repo/blob/CodeMirrorBlob.tsx [hunk 1]
```diff
@@ -220,6 +220,13 @@ export const Blob: React.FunctionComponent<BlobProps> = props => {
     })
     editorRef.current = editor
 
+    // Make gutters content visible for the screen readers
+    useEffect(() => {
+        if (editor) {
+            editor.dom.querySelector('.cm-gutters')?.removeAttribute('aria-hidden')
```
Comment (fkling) [inline client/web/src/repo/blob/CodeMirrorBlob.tsx:226]: This might not work anyways because the DOM is controlled by CodeMirror (so the attribute might be set again at some point).

**Comment by taras-yemets** [comment]
@philipp-spiess 

> I could navigate through the rows with ctrl+option+arrow right
Yes, but only on inside code rows; blame decorations don't seem to be available this way.

> Would it be possible to also put links into a screenreader-only widget?
I can hardly imagine a hidden clickable element. Maybe I'm missing some context here...

File: client/web/src/repo/blob/CodeMirrorBlob.tsx (hunk 1 above)
Comment (taras-yemets) [reply client/web/src/repo/blob/CodeMirrorBlob.tsx:226]: > This might not work anyways because the DOM is controlled by CodeMirror (so the attribute might be set again at some point).

What if we add gutters' dependencies (file blob, blame hunks, etc.) as `useEffect` dependencies? Could it help?
But yeah, it still sounds unsafe.

**Comment by felixfbecker** [comment]
Anything can be put into an `sr-only` element. It's simply moved off-screen. Which also makes it a bit hacky and a last-resort solution, because the VoiceOver cursor will be in awkward places instead of over the corresponding visual elements.

**Comment by felixfbecker** [comment]
It seems like a really bad issue for CodeMirror that Gutter decorations like blame and line numbers are completely inaccessible. GitHub's code view and old blob view exposed these. We should definitely at least report it upstream, I couldn't find any open issue from a quick search...

**Comment by felixfbecker** [comment]
On second thought, it's understandable why they do this. They want to use `contenteditable` for editing functionality. Obviously you wouldn't want the line numbers to be part of the editable area, so they are distinct subtrees of the DOM. So there isn't an easy solution I can think of that we could contribute upstream 🤔 

You really want the line numbers, gutter etc to be associated with each other somehow. We will have this problem with the diff view too, if we want to migrate to CodeMirror there – you want the rows of the left and right-hand side of the diff to be associated with each other to enable side-by-side and up-and-down traversal. That's why we used a `<table>` before (the diff view still does) and GitHub does too.

For line numbers, maybe the "association" could be achieved with `aria-describedby`, but for interactive content like blame and diff view it would get tricky.

The only idea I have is if CodeMirror used a table too, and maybe it would be possible to set `contenteditable="false"` on each line number and gutter cell, and add some JS handling to make sure <kbd>⏎</kbd> inserts a new line as a new row and not within the cell (they must have some kind of handling like this already for updating line numbers). I don't know if this would work though, and in the context of needing to support editing in CodeMirror, it would probably be quite a big change for us to make just to get good read-only accessibility support.

**Comment by fkling** [comment]
A couple of thoughts:

- It's possible to add "line decorations" to all (visible) lines and set arbitrary attributes on them (afaik). This could allow us to add the line number as `aria-*` attribute one way or another.
- How do screenreader users even want to consume additional information like git blame? I'm not a screenreader user so I can't answer that question, but one possibility would be to somehow associate keyboard bindings with aria announcements. E.g. pressing `gb` could read the blame information for the current line.
  More generally, maybe we need to design an overall different (better?) experience for screenreader users.
- I'll do some tests for my own understanding how CodeMirror works with screenreader and depending on those results I'll post on the discussion board about the issues raised here (e.g. line numbers). A11y is important to CodeMirror (for all I know). Even if some of these things cannot be changed on the CodeMirror side, Marijn might have suggestions for how to address these issues.

**Comment by philipp-spiess** [comment]
^ To add to the examples above:

**GitLab:** Similar to GitHub but slightly worse: The screenreader starts with the blame information, then reads all line numbers for the affect hunk, followed by the code:


https://user-images.githubusercontent.com/458591/201136521-83f77cc8-b630-4e2e-bc43-a191a3491b53.mov

**Comment by philipp-spiess** [comment]
I noticed another minor improvement that moving the git blame rendering inside a line could bring: Right now, when the viewport width is constrained, we resort to scrolling of the code view. If the git blame gutter is shown, it will take up a lot of space and make the scrollable area very small:

Screenshot 2022-11-10 at 17 06 38

**Comment by taras-yemets** [comment]
A couple of findings 👇🏻 

- Despite having `aria-hidden="true"` on the gutters wrapper, blame decorations are announced by the screen reader. Here's the video recorder on the `main` branch.

https://user-images.githubusercontent.com/25318659/201158388-e822abfb-d40d-496f-8dc8-f645a449d784.mov

- Adding `sr-only` blame decorations to the beginning of the code line (as suggested [here](https://github.com/sourcegraph/sourcegraph/pull/44148#issuecomment-1308970422)) helps with the order of reading: blame decoration content is announced before the line (code) content, but it has two drawbacks in my opinion:
  - Despite being `sr-only` 'hidden,' blame decoration is still focusable, and it results in having two popups. I think there is a way to tweak this behavior, but I'm not sure if it's worth the effort considering the next point.  
  - Having blame decoration text right before the line code content doesn't make the content more comprehensive, in my opinion. Blame decoration content and code content are announced as one big chunk of text (with mentioning blame decoration being a link, of course), and I can hardly understand the actual meaning. 

https://user-images.githubusercontent.com/25318659/201161873-c0085f25-5555-4a7e-ae64-375524d2800b.mov

- Moving blame decorations content from the gutter to [widget decorations](https://codemirror.net/docs/ref/#view.Decoration^widget) helps with the order of reading but still has the same comprehension problems as described in the previous point, in my opinion. 
Please ignore the styling problems in the video below; it's just for demo purposes.

https://user-images.githubusercontent.com/25318659/201163142-fc75e279-d928-4e17-993a-2a3e9205ae25.mov

After trying a bunch of different approaches, it seems to me that neither of them improves the screen reader's user experience dramatically, so I'm not sure if we should invest in the future development of any of them until we have a good understanding of what the ideal screen reader user journey is. Just to announce all the text content (line numbers, blame decorations, code content) without distinguishing it in a meaningful way seems insufficient (GitHub, GitLab, VSCode, and Google Code Search examples gathered by @philipp-spiess prove that, in my opinion).
As of now, we have blame decorations announced as `aria-hidden="true"` is ignored. Given that, do we still consider https://github.com/sourcegraph/sourcegraph/issues/43211 as critical for the upcoming audit?

cc: @philipp-spiess, @felixfbecker, @fkling

**Comment by philipp-spiess** [comment]
> Despite being sr-only 'hidden,' blame decoration is still focusable, and it results in having two popups. I think there is a way to tweak this behavior, but I'm not sure if it's worth the effort considering the next point.

I can only reproduce this with tab navigation, not with `ctrl+option+arrow right`. So in the normal reading order, this is not true _I believe_. I might be missing something though! 

> Having blame decoration text right before the line code content doesn't make the content more comprehensive, in my opinion. Blame decoration content and code content are announced as one big chunk of text (with mentioning blame decoration being a link, of course), and I can hardly understand the actual meaning.

Yeah one thing I noticed at both GitHub and GitLab is that they announce the line number between the blame text and the code text. I think we could do this to (sr_only). This could make up for the `aria-hidden` on the line number gutter.

**Comment by philipp-spiess** [comment]
> Having blame decoration text right before the line code content doesn't make the content more comprehensive, in my opinion. Blame decoration content and code content are announced as one big chunk of text (with mentioning blame decoration being a link, of course), and I can hardly understand the actual meaning.

I want to add here that It's probably hard for us to judge what is comprehensive and what not with how little experience (at least I) have with screen readers. There is still a difference though between connecting the blame information with the code line vs. not doing that. I would err on the side of giving _more_ data (in this case the connection to the line of code) over making the information inaccessible.

If this is distracting for users, they can always disable git blame anyways.

**Comment by taras-yemets** [comment]
> I can only reproduce this with tab navigation, not with ctrl+option+arrow right. So in the normal reading order, this is not true I believe. I might be missing something though!
More on that: https://accessibilityinsights.io/info-examples/web/aria-hidden-focus/

I totally missed that! Thank you, @philipp-spiess!

**Comment by taras-yemets** [comment]
> I want to add here that It's probably hard for us to judge what is comprehensive and what not with how little experience (at least I) have with screen readers.

Fair point, @philipp-spiess! 

> There is still a difference though between connecting the blame information with the code line vs. not doing that. I would err on the side of giving more data (in this case the connection to the line of code) over making the information inaccessible.

Agree. Will make another iteration on it.

**Comment by felixfbecker** [comment]
Thanks for all the investigation, these are great thoughts!

I think moving the blame into the code view as decorations is something we definitely should do, as that improves it from not accessible (only through <kbd>tab</kbd> is not enough, unfortunately) to at least accessible in some form. It may not be the most ergonomic in all use cases to have every blame line read out before every code line, but blame is toggleable, so it's not going to degrade the plain code reading experience and is better than not having blame accessible at all when the user explicitly asked for blame information. So I'd say if that is easy to do, we should go ahead and make that change.

Regarding comparing different blame views, I think GitHub's based on a `<table>` is the best, for the simple reason that it allows two-dimensional navigation. This gives users the ability to read the code and navigate the associated blame information in a flexible way without us having to educate them on all the keyboard shortcuts they would need for this first (we would be required by WCAG to explain these). The interaction is non-trivial considering blames can span multiple lines.

I recorded a quick video to demonstrate what I mean:

https://user-images.githubusercontent.com/10532611/201244596-96b80f1e-41d3-47c5-b0e4-2f7e15844059.mp4

I could be wrong on this too of course, I'm not a screen reader user either. And not every screen reader user is alike either. But I don't think we need to block changes on interviewing a screen reader user, we should be able to make reasonable decisions. This could also end up having a setting if someone prefers one or the other, but good defaults are also important.

Note that it doesn't have to literally be a `<table>` element. With moving it into the code area, the blame would I assume at least be in the same subtree, i.e. there is a wrapper element for each line, and it has children for both the blame and the code. If we can only control _attributes_ on those elements, we could give the line wrapper a `role="row"`, the blame and code `role="cell"` (and `aria-rowspan` as necessary), and the entire code view `role="table"`, and the screen reader would give us the rest for free.

If using `table` doesn't feel right semantically, consider that ARIA simply says that `table` is for any ["tabular content"](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/table_role) and the HTML spec says `table` ["represents data with more than one dimension"](https://html.spec.whatwg.org/multipage/tables.html#the-table-element). It's not tied to only being used for what one might think of a table as used in print. Tables got a bit of a bad rep due to their abuse in the 90s/early 2000s for purely presentational layout where the data is not actually two-dimensional (no semantic meaning of rows and columns) – which are now done by flexbox/CSS grid – but this is not one of those cases, the linkage between cells horizontally and vertically has semantic meaning here.
There's also `grid` in ARIA, which intuitively sounds like a better fit here, but `grid` is specifically for _interactive_ grids like those of a spreadsheet app. [When the content is static, the `table` role should be used](https://adrianroselli.com/2017/11/hey-its-still-ok-to-use-tables.html).

I would caveat against using editors/IDEs as a comparison because being in the editing context fundamentally changes constraints. There are some solutions that would be unfeasible if you're primarily targeting editing and have a cursor moving around – you're basically required to go through keyboard shortcuts to jump outside the "code" and editable area. On the other hand, it's also a more natural expectation for a user to learn the keyboard shortcuts for an IDE than for a web-based code reader. There are established patterns too in desktop apps on how to learn about these – every desktop app, whether macOS or Windows, has a menu bar with all (or most) actions that declare keyboard shortcuts, and a "Help" item with more references. All that is not established for websites; it's simply a different environment with different conventions. A desktop app wouldn't even naturally follow WCAG (the W is for Web after all) but different guidelines. Electron apps, of course, are a bit of a hybrid here but VS Code definitely leans more towards desktop app patterns, and it does e.g. have a desktop-style menu bar.

I would also caveat against going down the route of implementing duplicate bespoke DOM for screen readers only. @taras-yemets already pointed out some problems with that "duplication" and `sr-only`. It goes against providing _equal access_ to the same user interface and should be a last resort. Generally, accessibility support doesn't target a binary screen reader or non-screen reader user; it can be a spectrum where a screen reader user may use the screen reader, keyboard, and mouse in combination (e.g. maybe they have enough vision to click a rough area to jump to, or they ask someone else to help them when they get disoriented). And from a code structure perspective, what ends up happening is that the "accessible" mode erodes over time or breaks because it's being tested less. The best way to go is to have a unified DOM tree, where the same elements that are visible are also marked up semantically for keyboard and screen reader support.
Google goes the route of having a completely different "acccessibility mode" for Google Docs, because they decided on the constraint of a pixel-perfect consistent text rendering, and so they use a canvas and have to implement a complete shadow DOM just for accessibility purposes in addition to the canvas. But they also have a lot more money and frontend engineers than we do to deal with all the downsides and implications of that 🙂 

**TL;DR** I think the best intermediate step we can take would be to move blame into the code area, and I'd be curious to see if afterwards, we can customize the wrapper elements' attributes to try out enabling two-dimensional navigation.

**Comment by philipp-spiess** [comment]
100% agree, @felixfbecker. 

I think table role is new with the new GitHub UI as I only started noticing it last night (I only got access to the new UI last night) but it definitely make sense and would be a good follow-up improvement.

**Comment by philipp-spiess** [comment]
@taras-yemets Thanks for putting the PR up for review! 

I was checking it out locally and noticed a strange behavior while scrolling:

There seems to be a delay between the git blame being rendered and the code being rendered. In that the, the text is shifted and rendered where the git blame should be.

It also felt awkward when I first opened git blame and there was no UI feedback until the git blame data appeared from the backend.

Is there any chance we can always render at least an empty placeholder? 🤔  Or maybe at least make it so that the text always starts offset so that there's no layout shift?

https://user-images.githubusercontent.com/458591/202679799-5506fc7b-d0fc-4c36-acb6-8a3712151e0e.mov

**Comment by taras-yemets** [comment]
> There seems to be a delay between the git blame being rendered and the code being rendered. In that the, the text is shifted and rendered where the git blame should be.

@philipp-spiess, nice catch! Thank you! On it.

**Comment by taras-yemets** [comment]
> I was checking it out locally and noticed a strange behavior while scrolling:
There seems to be a delay between the git blame being rendered and the code being rendered. In that the, the text is shifted and rendered where the git blame should be.
It also felt awkward when I first opened git blame and there was no UI feedback until the git blame data appeared from the backend.
Is there any chance we can always render at least an empty placeholder? 🤔 Or maybe at least make it so that the text always starts offset so that there's no layout shift?

@philipp-spiess, addressed by rendering placeholders while waiting for the actual blame data.

https://user-images.githubusercontent.com/25318659/203364169-001d35ad-6dbf-4a6d-b7f2-3657fac77537.mov

File: client/web/src/repo/blob/codemirror/blame-decorations.tsx [hunk 2]
```diff
@@ -38,51 +47,30 @@ const [hoveredLine, setHoveredLine] = createUpdateableField<number | null>(null,
     }),
 ])
 
-/**
- * Used to find the blame decoration(s) with the longest text,
- * so that they can be used as gutter spacer.
- */
-const longestColumnDecorations = (hunks?: BlameHunk[]): BlameHunk | undefined =>
-    hunks?.reduce((acc, hunk) => {
```
Comment (philipp-spiess) [inline client/web/src/repo/blob/codemirror/blame-decorations.tsx:46]: Phew good that we got rid of this, this requires traversal through all hunks 😅

File: client/web/src/repo/blob/codemirror/blame-decorations.tsx [hunk 3]
```diff
@@ -117,32 +105,96 @@ class BlameDecoratorMarker extends GutterMarker {
...
                 }
-                return new BlameDecoratorMarker(view, hunk)
+
+                private computeDecorations(view: EditorView, facet: Facet<BlameHunk[], BlameHunk[]>): DecorationSet {
+                    const widgets = []
+                    const hunks = view.state.facet(facet)
+                    for (const { from, to } of view.visibleRanges) {
+                        for (let position = from; position <= to; ) {
+                            const line = view.state.doc.lineAt(position)
+                            const matchingHunk = hunks.find(hunk => hunk.startLine === line.number)
```
Comment (philipp-spiess) [inline client/web/src/repo/blob/codemirror/blame-decorations.tsx:131]: Should we convert `hunk` into a `Map` so we have faster lookup here in a follow-up? 

There's a catch though: With #44287 we'll need to be able to know what the hunk is _for every line_, not just the start line, so we can render the proper recency color.

So we'll need to do something like:

```js
const map = new Map<number, Hunk>();
for (const hunk of hunks) {
  for (const line = hunk.startLine; line <= hunk.endLine; line++) {
    if (map.has(line)) {
      console.error(
        "This shouldn't happen, backend sent overlapping blame hunks"
      );
    }
    map.set(line, hunk);
  }
}
```

Let's wait for this change though until we also merge #44199 which changes a lot on that infra.

**Review by philipp-spiess — APPROVED** [review]
LGTM! Obviously this is very hacky but the benefits are great and it makes the content accessible.

I posted some thoughts for future improvements but for now this looks good (I also played around with it locally and couldn't tell the difference)

**Comment by felixfbecker** [comment]
> Obviously this is very hacky but the benefits are great

@philipp-spiess could you elaborate the "obviously hacky" aspects and maybe file a follow-up issue to fix those? Since the whole CodeMirror migration is supposed to be a better more flexible architecture, we should of course strive to not have core features like this implemented in an "obviously hacky" way

**Comment by philipp-spiess** [comment]
@felixfbecker I was replying to the updated message in the PR description, to quote @taras-yemets:

> But it looks a bit hacky as we still use two different approaches (inline decorations and gutter) to achieve this effect and it may be not the way the CodeMirror authors expect us to use the editor. CodeMirror on its website claims that it "works well with screen readers and keyboard-only users" so I'll browse their forum to find similar use-cases or post an issue to get some guidance on how to make gutters keyboard and screenreader accessible. If there's a better approach we'll address it in a follow-up PR. cc: @fkling

I think the follow-up that Taras proposed sounds good as well. I think this is something that should be fixed upstream and maybe we can even help with that.
